### ISSUE TITLE
🔮 Use ES6 when we stop supporting IE entirely

### ISSUE THREAD

**Issue by chrisblakley** [chrisblakley/Nebula#1396]
This one might be a little while, but once we decide to stop supporting IE11 we can update a bunch of our JavaScript to ES6.

I'll include ideas of what to update here.

- [x] **Default parameters in function declarations!** https://caniuse.com/mdn-javascript_functions_default_parameters
- [x] Change `indexOf()` to `includes()` https://caniuse.com/es6-string-includes
  - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes
- [ ] Consider using arrow functions for quick, little things, but don't overuse them.
- [x] Named capture groups in RegEx! https://caniuse.com/mdn-javascript_builtins_regexp_named_capture_groups
  - [ ] Also lookbehinds in RegEx
- [x] Fetch API #1359
  - Consider replacing all AJAX functions with Fetch
  - Also use priority hints when appropriate on Fetch requests: #1817
- [x] `let` and `const` declarations. Use `let` for scoped variables, use `const` for immutable references.
- [x] Dynamic named keys in objects without repeating `objectName[dynamicKey] = 'whatever';` _(see below)_
- [x] JS Modules https://github.com/chrisblakley/Nebula/issues/1493
   - This one looks like it'll be dependent on more than just IE going away- some mobile browsers need to adopt support (Android, Samsung)
- [ ] Variable fonts #1710
- [ ] Constraint validation #1679
- [x] URL API (and URLsearchparams function for query strings) #1559
- [x] [async/await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) (and Promises)
- [x] Optional `catch` binding (don't need to use **(e)** every catch)
- [x] `for... of` loops
- [ ] `.some()` and `.every()` for checking array values https://caniuse.com/#feat=mdn-javascript_builtins_typedarray_some
- [x] Optional `catch` binding on try...catch https://caniuse.com/mdn-javascript_statements_try_catch_optional_catch_binding

Other features to integrate when IE11 support is dropped
- [x] Passive event listeners https://caniuse.com/#feat=passive-event-listener
- [x] Case insensitive attribute selectors https://caniuse.com/#feat=css-case-insensitive (not specifically ES6, but usable after IE11 support is dropped) ```a[href$='.pdf' i]```
- [x] Update to Mmenu v8 #1866 
- [ ] Update to Bootstrap v5 #1990 
- [ ] CSS Grid #1991 
- [ ] SVG Favicons
- [ ] `#rrggbbaa` syntax for hex color alpha support
- [ ] `rgb()` itself can accept alpha values (4th parameter) without needing `rgba()`
- [x] ~~`requestIdleCallback` https://caniuse.com/#feat=requestidlecallback~~ Still waiting on Safari!
- [ ] Consider adding Dark Mode support for Nebula defaults with `prefers-color-scheme`
- [ ] CSS `clamp()`: https://developer.mozilla.org/en-US/docs/Web/CSS/clamp https://twitter.com/argyleink/status/1310640086111846400

**Comment by chrisblakley** [comment]
Here's the info about the Fetch API moved from another issue about this same topic:

https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

Note: Not supported in IE at all (https://caniuse.com/#search=fetch)

Basic example:
```js
fetch('https://davidwalsh.name/some/url', {
	method: 'get'
}).then(function(response) {
	
}).catch(function(err) {
	// Error :(
});
```

More advanced example:
```js
var request = new Request('https://davidwalsh.name/users.json', {
	method: 'POST', 
	mode: 'cors', 
	redirect: 'follow',
	headers: new Headers({
		'Content-Type': 'text/plain'
	})
});

// Now use it!
fetch(request).then(function() { /* handle response */ });
```

Here is a good comparison: https://gist.github.com/hedleysmith/15fa60fda5ef4369636a4b23e018dc8f

**Comment by chrisblakley** [comment]
Note: IE11 support ends on October 14, 2025, so tuck this one away for a while...

**Comment by chrisblakley** [comment]
New way to allow dynamic named keys in an object:

```js
var key = 'Yolo'; //Imagine this changes or is pulled from some other source...

//Old way
var objectName = {};
objectName[key] = 'the value here';

//New way
var objectName = {
    [key]: 'the value here'
}
```

At the time of writing, here are the functions I found that use dynamic keys with the old method:
- `nvFormRealTime()`
- `nebulaTimer()`
- `addHTML5VideoPlayer()`
- `nebulaYoutubeReady()`
- `vimeoReady()`
- `addVimeoPlayer()`

Not all of those will need to be updated, but that's everything I could find.

**Comment by chrisblakley** [comment]
<img width="1321" alt="screen shot 2018-05-09 at 2 54 34 pm" src="https://user-images.githubusercontent.com/6349678/39833741-f7bb2696-5398-11e8-86e7-36f7c1bebff9.png">

Named capture groups in RegEx.

Also they recommend adding the flags `u` for unicode and `s` for dotAll (which makes the `.` actually match any character- no longer ignoring newlines).

**Comment by chrisblakley** [comment]
Lookaheads are not new:

<img width="1033" alt="screen shot 2018-05-09 at 2 58 13 pm" src="https://user-images.githubusercontent.com/6349678/39833932-7ee4f930-5399-11e8-8d0a-34da3d04447a.png">

Lookbehinds are new:

<img width="864" alt="screen shot 2018-05-09 at 2 59 09 pm" src="https://user-images.githubusercontent.com/6349678/39833948-8ae2ec10-5399-11e8-9501-aaa14383289e.png">

**Comment by chrisblakley** [comment]
Could Nebula take advantage of using proxy on objects for any reason?
https://www.atyantik.com/proxy-javascript-es6-feature/

**Comment by chrisblakley** [comment]
Just checked a few analytics properties and I'm seeing IE11 usage around 2.81% since March 2018 which is just about inline with where StatCounter is calculating (2.59%). Down to 2.63% for just August 2018.

**Comment by chrisblakley** [comment]
Use `do` to replace complex ternary operators:

![screen shot 2018-11-23 at 11 58 20 am](https://user-images.githubusercontent.com/6349678/48954368-2d69e100-ef17-11e8-8674-bccb0a7aeec6.png)

**Comment by chrisblakley** [comment]
Use `replaceAll()` instead of needing to use regex on `replace()`:
![screen shot 2018-11-23 at 12 01 03 pm](https://user-images.githubusercontent.com/6349678/48954455-85a0e300-ef17-11e8-9896-32699b62429a.png)

**Comment by chrisblakley** [comment]
The following are equivalent:
![screen shot 2018-11-23 at 12 02 44 pm](https://user-images.githubusercontent.com/6349678/48954496-b97c0880-ef17-11e8-992d-3730ca37ebe0.png)

**Comment by chrisblakley** [comment]
Get the last item from an array without needing to get the array length-1.

You can also easily change the last item of an array with it (Shown in the third line of this screenshot):
![screen shot 2018-11-23 at 12 04 39 pm](https://user-images.githubusercontent.com/6349678/48954552-0364ee80-ef18-11e8-8c81-290a5982c0af.png)

**Comment by chrisblakley** [comment]
Just noting here that it is rumored that Microsoft will be moving Edge to be a Chromium-based browser!

**Comment by chrisblakley** [comment]
Default parameters in functions:
![screen shot 2018-12-13 at 11 48 15 pm](https://user-images.githubusercontent.com/6349678/49983579-9d292580-ff31-11e8-8173-a0ab888968b6.png)

**Comment by chrisblakley** [comment]
`URLSearchParams()`
![screen shot 2018-12-13 at 11 54 56 pm](https://user-images.githubusercontent.com/6349678/49983759-95b64c00-ff32-11e8-8b3b-685c2fa54f39.png)

**Comment by chrisblakley** [comment]
Also, just to have it somewhere, this is also when we can fully utilize CSS variables too.

**Comment by chrisblakley** [comment]
Here's where and how I want to add passive event listeners:
https://github.com/chrisblakley/Nebula/issues/1323#issuecomment-461596025

**Comment by chrisblakley** [comment]
Just storing this [browser usage stat](https://caniuse.com/usage-table) here for historical reference (**Feb 7, 2019**):

![screen shot 2019-02-07 at 4 44 11 pm](https://user-images.githubusercontent.com/6349678/52444917-aed51680-2af7-11e9-8bbd-19ed01ecebaf.png)

I'll post a few more between now and October 2025 to pass the time...

**Comment by chrisblakley** [comment]
`fetch` in admin.js: https://github.com/chrisblakley/Nebula/issues/1821

**Comment by chrisblakley** [comment]
I want to research further usage of `import()` as a way of splitting up nebula.js (and main.js) into smaller chunks that get pulled in together similar to how style.scss works now.

**Comment by chrisblakley** [comment]
Use more priority hints as described here: https://github.com/chrisblakley/Nebula/issues/1817

**Comment by chrisblakley** [comment]
KV Storage with build-in modules too: https://github.com/chrisblakley/Nebula/issues/1832

**Comment by chrisblakley** [comment]
Just noting that Googlebot (search indexing) now uses the latest version ofChromium so all of this ES6 tech will be properly indexed for search.

**Comment by chrisblakley** [comment]
Use `fetch()` for autocomplete search too.

**Comment by chrisblakley** [comment]
I made a spreadsheet to chart the lowering market share of IE11.
https://docs.google.com/spreadsheets/d/1bTvbYZPwiQWf3HL5R1ABgSdvyUSvJvH1wy2gpgvLOuk/

**July 2019:**
![Screen Shot 2019-07-05 at 12 30 36 PM](https://user-images.githubusercontent.com/6349678/60735284-b969cf80-9f20-11e9-88bd-d107919d0754.png)

**Comment by chrisblakley** [comment]
Once officially using JS modules, be sure to preload the ones that are needed on every page (and don't preload any that are conditional– that would defeat the purpose).

https://developers.google.com/web/updates/2017/12/modulepreload

**Comment by chrisblakley** [comment]
Wait...! No! That's not how it's supposed to happen... Now IE11 will become the most popular browser by 2024!! God help us all!

![Screen Shot 2020-04-18 at 12 03 01 AM](https://user-images.githubusercontent.com/6349678/79627652-204f2280-8108-11ea-974c-4485a8d464e4.png)

**Comment by chrisblakley** [comment]
Don't forget to update `admin.js` to use modules like the Utilties module so we can stop duplicating functions and reduce some tech debt.

This will also be a great time to clean up the admin.js file itself! It's been a mess for a while now.

**Comment by chrisblakley** [comment]
Another note is that when we drop IE support entirely, we can remove this meta tag:
```html
<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1" />
```

**Comment by chrisblakley** [comment]
This can be implemented when we stop supporting IE11 as well: https://github.com/chrisblakley/Nebula/issues/1496

**Comment by chrisblakley** [comment]
LinkedIn will end support of IE11 as of July 1.

![EYCPqnaVAAAmebF](https://user-images.githubusercontent.com/6349678/82628649-642bc080-9bbb-11ea-9d1f-bd9d7d8ca12c.jpg)

**Comment by chrisblakley** [comment]
Don't forget to update `package.json` when IE11 is no longer supported.

**Comment by chrisblakley** [comment]
I had some issues with Reporting Observer even with support conditionals, but maybe once IE11 dies it would be a good time to try it again (even though Firefox and Safari do not yet support it).
https://caniuse.com/#feat=mdn-api_reportingobserver

**Comment by chrisblakley** [comment]
Could also replace jQuery `$.one()` with vanilla JS `once()`.
- https://caniuse.com/#feat=once-event-listener
- https://api.jquery.com/one/

**Comment by chrisblakley** [comment]
For reference, Bootstrap v5 Alpha 1 was released this week which drops support for IE10 and IE11.

**Comment by chrisblakley** [comment]
# Now waiting on IE11 to fall below 1% for both ES6 (this thread) and ES2020 (#1999).

**Comment by chrisblakley** [comment]
Adding `for... of` loops to this list. Good for looping through arrays without jQuery.
https://caniuse.com/#feat=mdn-javascript_statements_for_of

**Comment by chrisblakley** [comment]
Added a temporary label to the repo for issues awaiting end of IE11 support:
https://github.com/chrisblakley/Nebula/issues?q=is%3Aissue+is%3Aopen+label%3A%22%E2%98%A0%EF%B8%8F+Awaiting+End+of+IE11+Support%22

Obviously this issue has the bulk of them, but there are many others with much more specific information on each feature.

**Comment by chrisblakley** [comment]
Noting that IE11 usage in the US is now at 0.9% (still at 1.23% globally). I'd like to see both the US and global numbers drop below 1%, but we're getting very close now!

**Comment by chrisblakley** [comment]
I was originally not going to dive off the deep end with arrow functions, but just in case I decide to do it, here's the syntax. It would work perfectly with the way the `nebula` object is currently written, so it would not be difficult at all to do.

```
const printSum = (a, b) => {
  const result = a + b;
  console.log('Answer: ' + result);
  
  return result;
}
```

The above syntax would need to be only slightly modified to also include default parameters

**Comment by chrisblakley** [comment]
This also opens the idea of `prefers-color-scheme` media queries too.

https://caniuse.com/prefers-color-scheme
https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme

**Comment by chrisblakley** [comment]
Not that Nebula would use them out of the box, but Conical Gradients in CSS are supported once IE11 goes away too.
https://caniuse.com/css-conic-gradients

```css
border-image-source: conic-gradient(#5B22E8, #FF2362, #5B22E8);
```

https://codepen.io/argyleink/pen/pogZxaZ

![Screen Shot 2021-01-20 at 12 27 48 AM](https://user-images.githubusercontent.com/6349678/105131162-65005380-5ab6-11eb-8478-88f9681a620d.png)

**Comment by chrisblakley** [comment]
My _insurance_ company sent out a notification that they are no longer be supporting Internet Explorer. You **_know_** it's getting close when Insurance companies are sending notifications!

**Comment by chrisblakley** [comment]
# For the first time, the worldwide usage of IE11 is below 1% market share. In the US it is sitting at 1.33%.

**Comment by chrisblakley** [comment]
Just a note– all of the JavaScript documentation pages will be broken until their locations are updated to the new module filename.

**Comment by chrisblakley** [comment]
Closing this at this point. Most (all?) of the remaining tasks have their own issue threads.

**Comment by chrisblakley** [comment]
Just because I'm still following news on IE11 support– WordPress core will likely drop IE support with the next version (5.8 in July 2021) or the version after that (5.9 in December 2021).

**Comment by chrisblakley** [comment]
WordPress has officially announced that version 5.8 due in July 2021 will drop support for IE11:
https://wordpress.org/news/2021/05/dropping-support-for-internet-explorer-11/
